funflow/src/Control/FunFlow/ContentHashable.hs

```diff
@@ -439,3 +445,61 @@ instance ContentHashable IO DirectoryContent where
         flip contentHashUpdate (Path.dirname dir)
         >=> flip contentHashUpdate (DirectoryContent dir)
         $ ctx
+
+instance Monad m => ContentHashable m UTCTime where
+  contentHashUpdate ctx utcTime = let
+      secondsSinceEpoch = fromEnum . utcTimeToPOSIXSeconds $ utcTime
+    in contentHashUpdate ctx secondsSinceEpoch
+
+-- | Path to a file to be treated as _externally assured_.
+--
+--   An externally assured file is handled in a somewhat 'cheating' way by
+--   funflow. The 'ContentHashable' instance for such assumes that some external
+--   agent guarantees the integrity of the file being referenced. Thus, rather
+--   than hashing the file contents, we only consider its (absolute) path, size and
+--   modification time, which can be rapidly looked up from filesystem metadata.
+--
+--   For a similar approach, see the instance for 'ObjectInBucket' in
+--   Control.FunFlow.AWS.S3, where we exploit the fact that S3 is already
+--   content hashed to avoid performing any hashing.
+newtype ExternallyAssuredFile = ExternallyAssuredFile (Path.Path Path.Abs Path.File)
+  deriving (Generic, Show)
+
+instance Aeson.FromJSON ExternallyAssuredFile
+instance Aeson.ToJSON ExternallyAssuredFile
+instance Store ExternallyAssuredFile
+
+instance ContentHashable IO ExternallyAssuredFile where
+  contentHashUpdate ctx (ExternallyAssuredFile fp) = do
+    modTime <- Path.IO.getModificationTime fp
+    fSize <- fileSize <$> getFileStatus (Path.toFilePath fp)
+    flip contentHashUpdate fp
+      >=> flip contentHashUpdate modTime
+      >=> flip contentHashUpdate_storable fSize
+        $ ctx
+
+
+-- | Path to a directory to be treated as _externally assured_.
+--
+--   For an externally assured directory, we _do_ traverse its contents and verify
+--   those as we would externally assured files, rather than just relying on the
+--   directory path. Doing this traversal is pretty cheap, and it's quite likely
+--   for directory contents to be modified without modifying the contents.
+newtype ExternallyAssuredDirectory = ExternallyAssuredDirectory (Path.Path Path.Abs Path.Dir)
+  deriving (Generic, Show)
+
+instance Aeson.FromJSON ExternallyAssuredDirectory
+instance Aeson.ToJSON ExternallyAssuredDirectory
+instance Store ExternallyAssuredDirectory
+
+instance ContentHashable IO ExternallyAssuredDirectory where
+  contentHashUpdate ctx0 (ExternallyAssuredDirectory dir0) = do
+    -- Note that we don't bother looking at the relative directory paths and
+    -- including these in the hash. This is because the absolute hash gets
+    -- included every time we hash a file.
+    (dirs, files) <- Path.IO.listDir dir0
+    ctx' <- foldM hashFile ctx0 (sort files)
+    foldM hashDir ctx' (sort dirs)
+    where
+      hashFile ctx fp = contentHashUpdate ctx (ExternallyAssuredFile fp)
+      hashDir ctx dir = contentHashUpdate ctx (DirectoryContent dir)
```

Comment:
Shouldn't this be `ExternallyAssuredDirectory`?

Reply:
Whoops, yes